tools/xmlschema.py

```diff
@@ -0,0 +1,312 @@
+        lines.append(
+            "  <xsd:any minOccurs='0' maxOccurs='unbounded' processContents='lax'/>"
+        )
+        lines.append("</xsd:sequence>")
+    return lines
+
+
+def print_element(element: ElementTree.Element) -> List[str]:
+    """
+    Print a child element of the sdf definition
+    """
+    lines = []
+
+    elem_name = get_attribute(element, "name")
+    elem_type = get_attribute(element, "type")
+    elem_reqd = get_attribute(element, "required")
+
+    if elem_type and is_std_type(elem_type):
+        elem_type = xsd_type_string(elem_type)
+
+    if elem_reqd:
+        min_occurs, max_occurs = SDF_REQUIRED_TO_MIN_MAX_OCCURS[elem_reqd]
+        lines.append(f"<xsd:choice  minOccurs='{min_occurs}' maxOccurs='{max_occurs}'>")
+
+    if elem_type is None:
+        lines.append(f"<xsd:element name='{elem_name}'>")
+        lines.extend(indent_lines(print_documentation(element), 2))
+        lines.append("  <xsd:complexType>")
+        lines.append("    <xsd:choice maxOccurs='unbounded'>")
+
+        for child_element in element.findall("element"):
+            lines.extend(indent_lines(print_element(child_element), 6))
+
+        lines.append("    </xsd:choice>")
+
+        for attribute in element.findall("attribute"):
+            lines.extend(indent_lines(print_attribute(attribute), 4))
+
+        lines.append("  </xsd:complexType>")
+    else:
+        lines.append(f"<xsd:element name='{elem_name}' type='{elem_type}'>")
+        lines.extend(indent_lines(print_documentation(element), 2))
+
+    lines.append("</xsd:element>")
+    lines.append("</xsd:choice>")
+    return lines
+
+
+def print_attribute(element: ElementTree.Element) -> List[str]:
+    """
+    Print an attribute of the sdf definition
+    """
+    lines = []
+
+    elem_name = get_attribute(element, "name")
+    elem_type = get_attribute(element, "type")
+    elem_reqd = get_attribute(element, "required")
+    elem_default = get_attribute(element, "default")
+
+    if elem_type and is_std_type(elem_type):
+        elem_type = xsd_type_string(elem_type)
+
+    use = ""
+    default = ""
+
+    if elem_reqd == "1":
+        use = "use='required'"
+    elif elem_reqd == "0":
+        use = "use='optional'"
+        if elem_default is not None:
+            default = f"default='{elem_default}'"
+
+    lines.append(
+        f"<xsd:attribute name='{elem_name}' type='{elem_type}' {use} {default}>"
+    )
+    lines.extend(indent_lines(print_documentation(element), 2))
+    lines.append("</xsd:attribute>")
+    return lines
+
+
+def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
+    """
+    Print xsd for top level SDF element
+    """
+    lines = []
+
+    elem_name = get_attribute(element, "name")
+    elem_type = get_attribute(element, "type")
+
+    elements = element.findall("element")
+    attributes = element.findall("attribute")
+    includes = element.findall("include")
+
+    lines.extend(print_documentation(element))
```

Comment:
Should this be `elements` (plural) to match the ruby code?

Reply:
Not sure that I understand. The child elements are iterated below?

Reply:
Sorry, I was confused by the `elements` in the `_elem.elements["description"]` of the ruby code. Disregard.